Now using AEAD like AES128_GCM or ChaCha20_Poly1305

oxtoacart opened the pull request:
It works, but it seems very slow so I need to figure out what's going on there. CPU usage is very low, so it's not clear why it's so slow.

cc: @myleshorton @p3rcya

oxtoacart commented:
Ha, turns out the speed issue was my local WIFI network.

oxtoacart commented:
@myleshorton or @joesis Either of you feel like reviewing this one?

oxtoacart commented on session.go:
```diff
@@ -213,20 +248,25 @@ func (s *session) sendLoop(pingInterval time.Duration) {
 	coalesceLoop:
-		for coalescedBytes < coalesceThreshold {
+		// Coalesce as much as possible without exceeding maxSessionFrameSize
+		for startOfData+coalescedBytes+maxHMACSize+maxFrameSize < maxSessionFrameSize {
```
One nice side-benefit of adding the session-level framing is that we can coalesce even more aggressively and end up with fewer syscalls to TCP.

joesis replied:
So `maxHMACSize+maxFrameSize` is the maximum bytes of one encrypted frame right? So it depends on the `dataEncrypter` to use, though both AEAD implementations have an overhead of 16. It's better to be derived from the `cryptoSpec` when `startSession`.

oxtoacart replied:
Will fix.

joesis commented on session.go:
```diff
@@ -255,22 +295,38 @@ func (s *session) sendLoop(pingInterval time.Duration) {
+		coalescedBytes = len(encryptedFramesData)
+
+		// Add length header before data
+		lenBuf := sessionFrame[startOfData-2:]
```
Maybe use `lenSize` here.

oxtoacart replied:
Will do.

oxtoacart commented on crypto.go:
```diff
@@ -59,8 +140,8 @@ func newSecret(cipherCode Cipher) ([]byte, error) {
 
-func newIV(cipherCode Cipher) ([]byte, error) {
-	iv := make([]byte, cipherCode.ivSize())
+func newIV(size int) ([]byte, error) {
```
We have to use crypto/rand because these need to be cryptographically secure. Per comment from math/rand package, `For random numbers suitable for security-sensitive work, see the crypto/rand package.`

joesis replied:
Oh sorry, I mixed up the two!

oxtoacart commented on crypto.go:
```diff
@@ -27,30 +28,110 @@ func (c Cipher) secretSize() int {
+	return
+}
+
+func newSecret() ([]byte, error) {
```
Needs to use crypto/rand.

oxtoacart commented on crypto.go:
```diff
@@ -83,74 +164,137 @@ func newDecrypter(cipherCode Cipher, secret []byte, iv []byte) (func([]byte), er
-	sendIV, pt = consume(pt, ivSize)
-	recvIV, _ = consume(pt, ivSize)
+	cs = &cryptoSpec{}
+	cs.cipherCode = Cipher(_cipherCode[0])
```
Good idea.

oxtoacart commented on session.go:
```diff
@@ -23,8 +24,10 @@ type session struct {
 	net.Conn
 	windowSize    int
 	maxPadding    *big.Int
```
What's the benefit of storing it as an int? In the performance critical section, we use it as a big.Int, so might as well store it as such.

joesis replied:
I just felt it's unnecessary to convert the type to and from `big.Int`. `rand.Int` is called only when `needsPadding` (which is calculated from the int `maxPadding`) is true, and the cost of `big.NewInt` seems neglectable comparing to zeroing the padding memory.

oxtoacart commented on session.go:
```diff
@@ -58,130 +59,164 @@ func startSession(conn net.Conn, windowSize int, maxPadding int, pingInterval ti
 
 func (s *session) recvLoop() {
 	echoTS := make([]byte, tsSize)
+	lb := make([]byte, 2)
```
Will change.

oxtoacart commented on session.go:
```diff
@@ -58,130 +59,164 @@ func startSession(conn net.Conn, windowSize int, maxPadding int, pingInterval ti
+					s.onSessionError(err, nil)
+					return
+				}
+				ackedFrames := int(int32(binaryEncoding.Uint32(_ackedFrames)))
```
Will fix.

oxtoacart commented on session.go:
```diff
@@ -213,20 +248,25 @@ func (s *session) sendLoop(pingInterval time.Duration) {
 	lastPing := time.Now()
 
 	onFrame := func(frame []byte) {
```
Good idea, will do.

oxtoacart commented on lampshade.go:
```diff
@@ -74,74 +74,109 @@
 //
-//       Recv IV - initialization vector for messages from server -> client,
-//                 128 bits for AES_CTR, 96 bits for ChaCha20
+//       Send IV1/2 - 96 bits of initialization vector. IV1 is used for
```
I was just trying to keep the overall message format to within 80 lines for nice display in godoc.

oxtoacart commented on lampshade.go:
```diff
@@ -187,41 +222,45 @@ const (
 	winSize        = 4
 	maxPaddingSize = 1
```
Thanks!

oxtoacart commented on lampshade.go:
```diff
@@ -187,41 +222,45 @@ const (
 	winSize        = 4
 	maxPaddingSize = 1
 	tsSize         = 8
+	maxSecretSize  = 32
```
The actual secretSize depends on the specific cipher, which is why I named this maxSecretSize.

joesis replied:
Oh, I see. The init message has 32 bytes for secret, but how it's used depends on the cipher.

oxtoacart commented:
@joesis Thanks for the review! I think I've addressed all of your comments with my latest commit. I noticed that on low througput networks, latency was pretty bad, so I also made the coalescing less aggressive with the latest commit. That seems to have helped.

joesis commented:
@oxtoacart Code LGTM. Every corner is carefully designed to reuse memory if possible. Impressive!

oxtoacart commented:
Thanks for the review @joesis!
